Source/Core/Core/IOS/IOSC.cpp

```diff
@@ -0,0 +1,246 @@
+}
+
+ReturnCode IOSC::SetOwnership(Handle handle, u32 new_owner, u32 pid)
+{
+  if (!HasOwnership(handle, pid))
+    return IOSC_EACCES;
+
+  m_key_entries[handle].owner_mask = new_owner;
+  return IPC_SUCCESS;
+}
+
+void IOSC::LoadDefaultEntries()
+{
+  // TODO: add support for loading and writing to a BootMii / SEEPROM and OTP dump.
+
+  const EcWii& ec = EcWii::GetInstance();
+
+  const std::vector<u8> ecc_key{ec.GetNGPriv(), ec.GetNGPriv() + 30};
+  m_key_entries[HANDLE_CONSOLE_KEY] = {true, TYPE_SECRET_KEY, SUBTYPE_ECC233, std::move(ecc_key),
+                                       3};
+
+  // Unimplemented.
+  m_key_entries[HANDLE_CONSOLE_ID] = {true, TYPE_DATA, SUBTYPE_DATA, std::vector<u8>(4), 0xFFFFFFF};
+  m_key_entries[HANDLE_FS_KEY] = {true, TYPE_SECRET_KEY, SUBTYPE_AES128, std::vector<u8>(16), 5};
+  m_key_entries[HANDLE_FS_MAC] = {true, TYPE_SECRET_KEY, SUBTYPE_MAC, std::vector<u8>(20), 5};
+
+  m_key_entries[HANDLE_COMMON_KEY] = {true,
+                                      TYPE_SECRET_KEY,
+                                      SUBTYPE_AES128,
+                                      {{0xeb, 0xe4, 0x2a, 0x22, 0x5e, 0x85, 0x93, 0xe4, 0x48, 0xd9,
+                                        0xc5, 0x45, 0x73, 0x81, 0xaa, 0xf7}},
+                                      3};
+
+  // Unimplemented.
+  m_key_entries[HANDLE_PRNG_KEY] = {true, TYPE_SECRET_KEY, SUBTYPE_AES128, std::vector<u8>(16), 3};
+
+  m_key_entries[HANDLE_SD_KEY] = {true,
+                                  TYPE_SECRET_KEY,
+                                  SUBTYPE_AES128,
+                                  {{0xab, 0x01, 0xb9, 0xd8, 0xe1, 0x62, 0x2b, 0x08, 0xaf, 0xba,
+                                    0xd8, 0x4d, 0xbf, 0xc2, 0xa5, 0x5d}},
+                                  3};
+
+  // Unimplemented.
+  m_key_entries[HANDLE_BOOT2_VERSION] = {true, TYPE_DATA, SUBTYPE_VERSION, std::vector<u8>(4), 3};
+  m_key_entries[HANDLE_UNKNOWN_8] = {true, TYPE_DATA, SUBTYPE_VERSION, std::vector<u8>(4), 3};
+  m_key_entries[HANDLE_UNKNOWN_9] = {true, TYPE_DATA, SUBTYPE_VERSION, std::vector<u8>(4), 3};
+  m_key_entries[HANDLE_FS_VERSION] = {true, TYPE_DATA, SUBTYPE_VERSION, std::vector<u8>(4), 3};
+
+  m_key_entries[HANDLE_NEW_COMMON_KEY] = {true,
+                                          TYPE_SECRET_KEY,
+                                          SUBTYPE_AES128,
+                                          {{0x63, 0xb8, 0x2b, 0xb4, 0xf4, 0x61, 0x4e, 0x2e, 0x13,
+                                            0xf2, 0xfe, 0xfb, 0xba, 0x4c, 0x9b, 0x7e}},
+                                          3};
+}
+
+IOSC::KeyEntries::iterator IOSC::FindFreeEntry()
+{
+  return std::find_if(m_key_entries.begin(), m_key_entries.end(),
+                      [](const auto& entry) { return !entry.in_use; });
+}
+
+IOSC::Handle IOSC::GetHandleFromIterator(IOSC::KeyEntries::iterator iterator) const
+{
+  _assert_(iterator != m_key_entries.end());
+  return iterator - m_key_entries.begin();
```

Comment:
This expression returns the same data type that `std::distance` would in this case. (pointer subtraction returns a `ptrdiff_t`). So either way a cast is technically needed.